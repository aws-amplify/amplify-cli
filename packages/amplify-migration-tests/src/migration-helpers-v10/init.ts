import { addCircleCITags, getCLIPath, nspawn as spawn } from '@aws-amplify/amplify-e2e-core';
import { EOL } from 'os';

const defaultSettings = {
  name: EOL,
  // eslint-disable-next-line spellcheck/spell-checker
  envName: 'integtest',
  editor: EOL,
  appType: EOL,
  framework: EOL,
  srcDir: EOL,
  distDir: EOL,
  buildCmd: EOL,
  startCmd: EOL,
  useProfile: EOL,
  profileName: EOL,
  region: process.env.CLI_REGION,
  local: false,
  disableAmplifyAppCreation: true,
  disableCIDetection: false,
  providerConfig: undefined,
  permissionsBoundaryArn: undefined,
};

export function initJSProjectWithProfileV10(cwd: string, settings?: Partial<typeof defaultSettings>): Promise<void> {
  const mergedSettings = { ...defaultSettings, ...settings };
  let env;

  if (mergedSettings.disableAmplifyAppCreation === true) {
    env = {
      CLI_DEV_INTERNAL_DISABLE_AMPLIFY_APP_CREATION: '1',
    };
  }

  addCircleCITags(cwd);

  const cliArgs = ['init'];
  const providerConfigSpecified = !!mergedSettings.providerConfig && typeof mergedSettings.providerConfig === 'object';
  if (providerConfigSpecified) {
    cliArgs.push('--providers', JSON.stringify(mergedSettings.providerConfig));
  }

  if (mergedSettings.permissionsBoundaryArn) {
    cliArgs.push('--permissions-boundary', mergedSettings.permissionsBoundaryArn);
  }

  if (mergedSettings?.name?.length > 20)
    console.warn('Project names should not be longer than 20 characters. This may cause tests to break.');

  return new Promise((resolve, reject) => {
    const chain = spawn(getCLIPath(), cliArgs, {
      cwd,
      stripColors: true,
      env,
      disableCIDetection: mergedSettings.disableCIDetection,
    })
      .wait('Enter a name for the project')
      .sendLine(mergedSettings.name)
      .wait('Initialize the project with the above configuration?')
      .sendConfirmNo()
      .wait('Enter a name for the environment')
      .sendLine(mergedSettings.envName)
      .wait('Choose your default editor:')
      .sendLine(mergedSettings.editor)
      .wait("Choose the type of app that you're building")
      .sendLine(mergedSettings.appType)
      .wait('What javascript framework are you using')
      .sendLine(mergedSettings.framework)
      .wait('Source Directory Path:')
      .sendLine(mergedSettings.srcDir)
      .wait('Distribution Directory Path:')
      .sendLine(mergedSettings.distDir)
      .wait('Build Command:')
      .sendLine(mergedSettings.buildCmd)
      .wait('Start Command:')
      .sendCarriageReturn();

    if (!providerConfigSpecified) {
      chain
        .wait('Using default provider  awscloudformation')
        .wait('Select the authentication method you want to use:')
        .sendCarriageReturn()
        .wait('Please choose the profile you want to use')
        .sendLine(mergedSettings.profileName);
    }
    chain
      .wait('Help improve Amplify CLI by sharing non sensitive configurations on failures')
      .sendYes()
      .wait(/Try "amplify add api" to create a backend API and then "amplify (push|publish)" to deploy everything/)
      .run((err: Error) => {
        if (err) {
          reject(err);
        } else {
          resolve();
        }
      });
  });
}
