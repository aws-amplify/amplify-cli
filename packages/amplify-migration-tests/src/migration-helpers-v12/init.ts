import { getCLIPath, nspawn as spawn } from '@aws-amplify/amplify-e2e-core';
import { EOL } from 'os';

const defaultSettings = {
  name: EOL,
  // eslint-disable-next-line spellcheck/spell-checker
  envName: 'integtest',
  editor: EOL,
  appType: EOL,
  framework: EOL,
  srcDir: EOL,
  distDir: EOL,
  buildCmd: EOL,
  startCmd: EOL,
  useProfile: EOL,
  profileName: EOL,
  region: process.env.CLI_REGION,
  local: false,
  disableAmplifyAppCreation: true,
  disableCIDetection: false,
  providerConfig: undefined,
  permissionsBoundaryArn: undefined,
};

export function initJSProjectWithProfileV12(cwd: string, settings?: Partial<typeof defaultSettings>): Promise<void> {
  const s = { ...defaultSettings, ...settings };
  let env;

  if (s.disableAmplifyAppCreation === true) {
    env = {
      CLI_DEV_INTERNAL_DISABLE_AMPLIFY_APP_CREATION: '1',
    };
  }

  const cliArgs = ['init'];
  const providerConfigSpecified = !!s.providerConfig && typeof s.providerConfig === 'object';
  if (providerConfigSpecified) {
    cliArgs.push('--providers', JSON.stringify(s.providerConfig));
  }

  if (s.permissionsBoundaryArn) {
    cliArgs.push('--permissions-boundary', s.permissionsBoundaryArn);
  }

  if (s?.name?.length > 20) console.warn('Project names should not be longer than 20 characters. This may cause tests to break.');

  const chain = spawn(getCLIPath(), cliArgs, {
    cwd,
    stripColors: true,
    env,
    disableCIDetection: s.disableCIDetection,
  })
    .wait('Enter a name for the project')
    .sendLine(s.name)
    .wait('Initialize the project with the above configuration?')
    .sendConfirmNo()
    .wait('Enter a name for the environment')
    .sendLine(s.envName)
    .wait('Choose your default editor:')
    .sendLine(s.editor)
    .wait("Choose the type of app that you're building")
    .sendLine(s.appType)
    .wait('What javascript framework are you using')
    .sendLine(s.framework)
    .wait('Source Directory Path:')
    .sendLine(s.srcDir)
    .wait('Distribution Directory Path:')
    .sendLine(s.distDir)
    .wait('Build Command:')
    .sendLine(s.buildCmd)
    .wait('Start Command:')
    .sendCarriageReturn();

  if (!providerConfigSpecified) {
    chain
      .wait('Using default provider  awscloudformation')
      .wait('Select the authentication method you want to use:')
      .sendCarriageReturn()
      .wait('Please choose the profile you want to use')
      .sendLine(s.profileName);
  }
  return chain
    .wait(/Help improve Amplify CLI by sharing non( |-)sensitive( | project )configurations on failures/)
    .sendYes()
    .wait(/Try "amplify add api" to create a backend API and then "amplify (push|publish)" to deploy everything/)
    .runAsync();
}

export function initIosProjectWithProfileV12(cwd: string, settings: Record<string, unknown>): Promise<void> {
  const s = { ...defaultSettings, ...settings };

  let env;

  if (s.disableAmplifyAppCreation === true) {
    env = {
      CLI_DEV_INTERNAL_DISABLE_AMPLIFY_APP_CREATION: '1',
    };
  }

  return spawn(getCLIPath(), ['init'], {
    cwd,
    stripColors: true,
    env,
  })
    .wait('Do you want to continue with Amplify Gen 1?')
    .sendYes()
    .wait('Why would you like to use Amplify Gen 1?')
    .sendCarriageReturn()
    .wait('Enter a name for the project')
    .sendLine(s.name)
    .wait('Initialize the project with the above configuration?')
    .sendConfirmNo()
    .wait('Enter a name for the environment')
    .sendLine(s.envName)
    .wait('Choose your default editor:')
    .sendLine(s.editor)
    .wait("Choose the type of app that you're building")
    .sendKeyDown(3)
    .sendCarriageReturn()
    .wait('Select the authentication method you want to use:')
    .sendCarriageReturn()
    .wait('Please choose the profile you want to use')
    .sendLine(s.profileName)
    .wait(/Help improve Amplify CLI by sharing non( |-)sensitive( | project )configurations on failures/)
    .sendYes()
    .wait(/Try "amplify add api" to create a backend API and then "amplify (push|publish)" to deploy everything/)
    .runAsync();
}

export function initAndroidProjectWithProfileV12(cwd: string, settings: Partial<typeof defaultSettings>): Promise<void> {
  const s = { ...defaultSettings, ...settings };

  let env;

  if (s.disableAmplifyAppCreation) {
    env = {
      CLI_DEV_INTERNAL_DISABLE_AMPLIFY_APP_CREATION: '1',
    };
  }

  return spawn(getCLIPath(), ['init'], {
    cwd,
    stripColors: true,
    env,
  })
    .wait('Enter a name for the project')
    .sendLine(s.name)
    .wait('Initialize the project with the above configuration?')
    .sendConfirmNo()
    .wait('Enter a name for the environment')
    .sendLine(s.envName)
    .wait('Choose your default editor:')
    .sendLine(s.editor)
    .wait("Choose the type of app that you're building")
    .sendKeyDown(1)
    .sendCarriageReturn()
    .wait('Where is your Res directory')
    .sendCarriageReturn()
    .wait('Select the authentication method you want to use:')
    .sendCarriageReturn()
    .wait('Please choose the profile you want to use')
    .sendLine(s.profileName)
    .wait(/Help improve Amplify CLI by sharing non( |-)sensitive( | project )configurations on failures/)
    .sendYes()
    .wait(/Try "amplify add api" to create a backend API and then "amplify (push|publish)" to deploy everything/)
    .runAsync();
}
