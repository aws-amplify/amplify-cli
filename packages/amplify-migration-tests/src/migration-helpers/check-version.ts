import { getCLIPath, nspawn as spawn } from 'amplify-e2e-core';

export function versionCheck(cwd: string, testingWithLatestCodebase = false, version: { v?: string } = {}): Promise<void> {
  return new Promise((resolve, reject) => {
    spawn(getCLIPath(testingWithLatestCodebase), ['-v'], { cwd, stripColors: true })
      .wait(/\d+\.\d+\.\d+/, v => (version.v = v.trim()))
      .run((err: Error) => {
        if (!err) {
          resolve();
        } else {
          reject(err);
        }
      });
  });
}

/**
 * This list is used to check migration tests with the following changes. (excludes layer migration tests)
 *
 * api add/update flow: https://github.com/aws-amplify/amplify-cli/pull/8287
 *
 * ext migrate flow: https://github.com/aws-amplify/amplify-cli/pull/8806
 */
export const allowedVersionsToMigrateFrom = ['5.2.0', '6.1.0'];
