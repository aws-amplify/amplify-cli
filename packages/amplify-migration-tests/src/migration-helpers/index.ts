export * from './init';
export * from './api';
export * from './check-version';
export * from './legacy-lambda-layer';
