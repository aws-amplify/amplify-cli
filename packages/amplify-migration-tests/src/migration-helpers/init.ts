import { addCircleCITags, getCLIPath, nspawn as spawn } from 'amplify-e2e-core';
import { EOL } from 'os';

const defaultSettings = {
  name: EOL,
  envName: 'integtest',
  editor: EOL,
  appType: EOL,
  framework: EOL,
  srcDir: EOL,
  distDir: EOL,
  buildCmd: EOL,
  startCmd: EOL,
  useProfile: EOL,
  profileName: EOL,
  region: process.env.CLI_REGION,
  local: false,
  disableAmplifyAppCreation: true,
};

export function initJSProjectWithProfileOldDX(cwd: string, settings: Object, testingWithLatestCodebase = false): Promise<void> {
  const s = { ...defaultSettings, ...settings };
  let env;

  if (s.disableAmplifyAppCreation === true) {
    env = {
      CLI_DEV_INTERNAL_DISABLE_AMPLIFY_APP_CREATION: '1',
    };
  }

  addCircleCITags(cwd);

  return new Promise((resolve, reject) => {
    spawn(getCLIPath(testingWithLatestCodebase), ['init'], { cwd, stripColors: true, env })
      .wait('Enter a name for the project')
      .sendLine(s.name)
      .wait('Enter a name for the environment')
      .sendLine(s.envName)
      .wait('Choose your default editor:')
      .sendLine(s.editor)
      .wait("Choose the type of app that you're building")
      .sendLine(s.appType)
      .wait('What javascript framework are you using')
      .sendLine(s.framework)
      .wait('Source Directory Path:')
      .sendLine(s.srcDir)
      .wait('Distribution Directory Path:')
      .sendLine(s.distDir)
      .wait('Build Command:')
      .sendLine(s.buildCmd)
      .wait('Start Command:')
      .sendCarriageReturn()
      .wait('Using default provider  awscloudformation')
      .wait(/(Select the authentication method you want to use|Do you want to use an AWS profile)/)
      .sendCarriageReturn()
      .wait('Please choose the profile you want to use')
      .sendLine(s.profileName)
      .wait(/Try "amplify add api" to create a backend API and then "amplify (push|publish)" to deploy everything/)
      .run((err: Error) => {
        if (!err) {
          resolve();
        } else {
          reject(err);
        }
      });
  });
}

export function initJSProjectWithProfile(cwd: string, settings: Object, testingWithLatestCodebase = false): Promise<void> {
  const s = { ...defaultSettings, ...settings };
  let env;

  if (s.disableAmplifyAppCreation === true) {
    env = {
      CLI_DEV_INTERNAL_DISABLE_AMPLIFY_APP_CREATION: '1',
    };
  }

  addCircleCITags(cwd);

  return new Promise((resolve, reject) => {
    spawn(getCLIPath(testingWithLatestCodebase), ['init'], { cwd, stripColors: true, env })
      .wait('Enter a name for the project')
      .sendLine(s.name)
      .wait('Initialize the project with the above configuration?')
      .sendConfirmNo()
      .wait('Enter a name for the environment')
      .sendLine(s.envName)
      .wait('Choose your default editor:')
      .sendLine(s.editor)
      .wait("Choose the type of app that you're building")
      .sendLine(s.appType)
      .wait('What javascript framework are you using')
      .sendLine(s.framework)
      .wait('Source Directory Path:')
      .sendLine(s.srcDir)
      .wait('Distribution Directory Path:')
      .sendLine(s.distDir)
      .wait('Build Command:')
      .sendLine(s.buildCmd)
      .wait('Start Command:')
      .sendCarriageReturn()
      .wait('Using default provider  awscloudformation')
      .wait(/(Select the authentication method you want to use|Do you want to use an AWS profile)/)
      .sendCarriageReturn()
      .wait('Please choose the profile you want to use')
      .sendLine(s.profileName)
      .wait(/Try "amplify add api" to create a backend API and then "amplify (push|publish)" to deploy everything/)
      .run((err: Error) => {
        if (!err) {
          resolve();
        } else {
          reject(err);
        }
      });
  });
}
