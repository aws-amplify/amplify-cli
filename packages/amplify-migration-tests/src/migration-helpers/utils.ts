import {
  amplifyPull,
  amplifyPushForce,
  amplifyPushWithoutCodegen,
  cliInputsExists,
  createNewProjectDir,
  deleteProjectDir,
  getAppId,
  getBackendConfig,
  getCLIInputs,
  getCloudFormationTemplate,
  getParameters,
  parametersExists,
} from '@aws-amplify/amplify-e2e-core';
import * as cfnDiff from '@aws-cdk/cloudformation-diff';
import { Writable } from 'stream';
import { AmplifyCategories } from '@aws-amplify/amplify-cli-core';
import strip from 'strip-ansi';
import { v4 as uuid } from 'uuid';

/**
 * generates a random string
 */
export const getShortId = (): string => {
  const [shortId] = uuid().split('-');

  return shortId;
};

/**
 * Given the parameter objects of each project for the provided category & resource key,
 * you can modify the parameter objects before the diff is performed.
 *
 * You can use conditional logic to delete attributes on the parameter objects before they are diffed, if
 * you want to exclude those attributes from the comparison. Make sure to return the modified objects
 * that you want to be diffed.
 */
export type ExcludeFromParameterDiff = (
  currentCategory: string,
  currentResourceKey: string,
  parameters: {
    project1: Record<string, unknown>;
    project2: Record<string, unknown>;
  },
) => { project1: Record<string, unknown>; project2: Record<string, unknown> };

/**
 * Asserts that parameters between two project directories didn't drift.
 */
export const assertNoParameterChangesBetweenProjects = (
  projectRoot1: string,
  projectRoot2: string,
  options?: {
    excludeFromParameterDiff?: ExcludeFromParameterDiff;
  },
): void => {
  const backendConfig1 = getBackendConfig(projectRoot1);
  const backendConfig2 = getBackendConfig(projectRoot2);
  expect(backendConfig2).toMatchObject(backendConfig1);
  for (const categoryKey of Object.keys(backendConfig1)) {
    const category = backendConfig1[categoryKey];
    if (Object.values(AmplifyCategories).includes(categoryKey)) {
      for (const resourceKey of Object.keys(category)) {
        if (cliInputsExists(projectRoot1, categoryKey, resourceKey)) {
          const cliInputs1 = getCLIInputs(projectRoot1, categoryKey, resourceKey);
          const cliInputs2 = getCLIInputs(projectRoot2, categoryKey, resourceKey);
          expect(cliInputs1).toEqual(cliInputs2);
        }
        if (parametersExists(projectRoot1, categoryKey, resourceKey)) {
          let parameters1 = getParameters(projectRoot1, categoryKey, resourceKey);
          let parameters2 = getParameters(projectRoot2, categoryKey, resourceKey);
          if (options && options.excludeFromParameterDiff) {
            const afterExclusions = options.excludeFromParameterDiff(categoryKey, resourceKey, {
              project1: parameters1,
              project2: parameters2,
            });
            parameters1 = afterExclusions.project1;
            parameters2 = afterExclusions.project2;
          }
          expect(parameters1).toEqual(parameters2);
        }
      }
    }
  }
};

class InMemoryWritable extends Writable {
  private _payload = '';

  _write(chunk: unknown, __encoding: BufferEncoding, callback: (error?: Error | null) => void): void {
    if (chunk) {
      this._payload += chunk.toString();
    }
    callback();
  }

  toString(): string {
    return this._payload;
  }
}

/**
 * Given the CFN templates of each project for the provided category & resource key,
 * you can modify the CFN templates objects before the diff is performed.
 *
 * You can use conditional logic to delete attributes on the CFN object before they are diffed, if
 * you want to exclude those attributes from the comparison. Make sure to return the modified objects
 * that you want to be diffed.
 */
export type ExcludeFromCFNDiff = (
  currentCategory: string,
  currentResourceKey: string,
  cfnTemplates: {
    project1: Record<string, unknown>;
    project2: Record<string, unknown>;
  },
) => { project1: Record<string, unknown>; project2: Record<string, unknown> };

/**
 * Collects all differences between cloud formation templates into a single string.
 */
export const collectCloudformationDiffBetweenProjects = (
  projectRoot1: string,
  projectRoot2: string,
  excludeFn?: ExcludeFromCFNDiff,
): string => {
  const backendConfig1 = getBackendConfig(projectRoot1);
  const backendConfig2 = getBackendConfig(projectRoot2);
  expect(backendConfig2).toMatchObject(backendConfig1);
  const stream = new InMemoryWritable();
  for (const categoryKey of Object.keys(backendConfig1)) {
    const category = backendConfig1[categoryKey];
    for (const resourceKey of Object.keys(category)) {
      let template1 = getCloudFormationTemplate(projectRoot1, categoryKey, resourceKey);
      let template2 = getCloudFormationTemplate(projectRoot2, categoryKey, resourceKey);

      // Description does not matter much and it can contain os/runtime specific words.
      delete template1.Description;
      delete template2.Description;

      if (excludeFn) {
        const afterExclusions = excludeFn(categoryKey, resourceKey, { project1: template1, project2: template2 });
        template1 = afterExclusions.project1;
        template2 = afterExclusions.project2;
      }

      const templateDiff = cfnDiff.diffTemplate(template1, template2);
      if (!templateDiff.isEmpty) {
        cfnDiff.formatDifferences(stream, templateDiff);
      }
    }
  }
  return strip(stream.toString());
};

/**
 * Pulls and pushes project with latest codebase. Validates parameter and cfn drift.
 */
export const pullPushWithLatestCodebaseValidateParameterAndCfnDrift = async (projRoot: string, projName: string): Promise<void> => {
  const appId = getAppId(projRoot);
  expect(appId).toBeDefined();
  const projRoot2 = await createNewProjectDir(`${projName}2`);
  try {
    await amplifyPull(projRoot2, { emptyDir: true, appId }, true);
    assertNoParameterChangesBetweenProjects(projRoot, projRoot2);
    expect(collectCloudformationDiffBetweenProjects(projRoot, projRoot2)).toMatchSnapshot();
    await amplifyPushWithoutCodegen(projRoot2, true);
    assertNoParameterChangesBetweenProjects(projRoot, projRoot2);
    expect(collectCloudformationDiffBetweenProjects(projRoot, projRoot2)).toMatchSnapshot();
  } finally {
    deleteProjectDir(projRoot2);
  }
};

/**
 * Pulls and pushes with force project with latest codebase. Validates parameter and cfn drift.
 */
export const pullPushForceWithLatestCodebaseValidateParameterAndCfnDrift = async (projRoot: string, projRoot2: string): Promise<void> => {
  const appId = getAppId(projRoot);
  expect(appId).toBeDefined();
  try {
    await amplifyPull(projRoot2, { emptyDir: true, appId }, true);
    assertNoParameterChangesBetweenProjects(projRoot, projRoot2);
    expect(collectCloudformationDiffBetweenProjects(projRoot, projRoot2)).toMatchSnapshot();
    await amplifyPushForce(projRoot2, true);
    assertNoParameterChangesBetweenProjects(projRoot, projRoot2);
    expect(collectCloudformationDiffBetweenProjects(projRoot, projRoot2)).toMatchSnapshot();
  } finally {
    deleteProjectDir(projRoot2);
  }
};
