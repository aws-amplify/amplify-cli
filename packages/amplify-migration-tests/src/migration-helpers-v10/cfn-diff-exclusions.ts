import { ExcludeFromCFNDiff } from '../migration-helpers/utils';

/**
 * Due to a known limitation of APIGateway, we do not want the AWS::ApiGateway::Deployment resource
 * from being compared in the CFN diff because it is regenerated whenever an amplify app is pulled down.
 * This would produce a false positive when checking for differences in the CFN templates of project1 & project2.
 * https://github.com/aws/aws-cdk/issues/8646#issuecomment-647561856
 *
 * AWS::ApiGateway::GatewayResponse is also excluded
 */
export const cfnDiffExclusions: ExcludeFromCFNDiff = (
  currentCategory: string,
  currentResourceKey: string,
  cfnTemplates: {
    project1: Record<string, unknown>;
    project2: Record<string, unknown>;
  },
) => {
  const excludeAPIGateWayDeploymentResource = (cfnTemplate: Record<string, unknown>): void => {
    const resources = cfnTemplate.Resources ?? {};
    const resourceKeys = Object.keys(resources);
    for (const key of resourceKeys) {
      const resource = resources[key];
      if (resource.Type === 'AWS::ApiGateway::Deployment' || resource.Type === 'AWS::ApiGateway::GatewayResponse') {
        delete resources[key];
      }
      if (resource.Type === 'AWS::AppSync::ApiKey' && resource.Properties) {
        delete resource.Properties.Expires;
      }
    }
  };
  if (currentCategory === 'api') {
    excludeAPIGateWayDeploymentResource(cfnTemplates.project1);
    excludeAPIGateWayDeploymentResource(cfnTemplates.project2);
  }
  return { project1: cfnTemplates.project1, project2: cfnTemplates.project2 };
};
