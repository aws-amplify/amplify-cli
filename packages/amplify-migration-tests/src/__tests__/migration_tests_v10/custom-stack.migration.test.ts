import { JSONUtilities } from '@aws-amplify/amplify-cli-core';
import {
  addCDKCustomResource,
  addCFNCustomResource,
  amplifyPull,
  amplifyPushAuth,
  amplifyPushAuthV10,
  buildCustomResources,
  createNewProjectDir,
  deleteProject,
  deleteProjectDir,
  getAppId,
  getProjectMeta,
} from '@aws-amplify/amplify-e2e-core';
import * as fs from 'fs-extra';
import * as path from 'path';
import { v4 as uuid } from 'uuid';
import { initJSProjectWithProfileV10 } from '../../migration-helpers-v10/init';
import { assertNoParameterChangesBetweenProjects, collectCloudformationDiffBetweenProjects } from '../../migration-helpers/utils';

describe('adding custom resources migration test', () => {
  let projRoot: string;
  beforeEach(async () => {
    projRoot = await createNewProjectDir('custom-resources');
  });

  afterEach(async () => {
    await deleteProject(projRoot, null, true);
    deleteProjectDir(projRoot);
  });

  it('add/update CDK and CFN custom resources', async () => {
    const cdkResourceName = `custom${uuid().split('-')[0]}`;
    const cfnResourceName = `custom${uuid().split('-')[0]}`;
    const cfnResourceNameWithV10 = `custom${uuid().split('-')[0]}`;

    await initJSProjectWithProfileV10(projRoot, { name: 'customMigration', disableAmplifyAppCreation: false });
    const appId = getAppId(projRoot);
    expect(appId).toBeDefined();

    await addCDKCustomResource(projRoot, { name: cdkResourceName });
    await addCFNCustomResource(projRoot, { name: cfnResourceNameWithV10, promptForCategorySelection: true });
    const srcCFNCustomResourceFilePath = path.join(__dirname, '..', '..', '..', 'custom-resources', 'custom-cfn-stack.json');
    // adding a resource to custom cfn stack
    const destCFNCustomResourceFilePath = path.join(
      projRoot,
      'amplify',
      'backend',
      'custom',
      cfnResourceNameWithV10,
      `${cfnResourceNameWithV10}-cloudformation-template.json`,
    );
    fs.copyFileSync(srcCFNCustomResourceFilePath, destCFNCustomResourceFilePath);

    // this is where we will write our custom cdk stack logic to
    const destCustomResourceFilePath = path.join(projRoot, 'amplify', 'backend', 'custom', cdkResourceName, 'cdk-stack.ts');
    const cfnFilePath = path.join(
      projRoot,
      'amplify',
      'backend',
      'custom',
      cdkResourceName,
      'build',
      `${cdkResourceName}-cloudformation-template.json`,
    );
    const srcCustomResourceFilePath = path.join(__dirname, '..', '..', '..', 'custom-resources', 'custom-cdk-stack-v10.ts');
    fs.copyFileSync(srcCustomResourceFilePath, destCustomResourceFilePath);
    await buildCustomResources(projRoot);
    await amplifyPushAuthV10(projRoot);

    // check if cfn file is generated in the build dir
    expect(fs.existsSync(cfnFilePath)).toEqual(true);
    const buildCFNFileJSON: any = JSONUtilities.readJson(cfnFilePath);
    // Basic sanity generated CFN file content check
    expect(buildCFNFileJSON?.Parameters).toEqual({
      env: { Type: 'String', Description: 'Current Amplify CLI env name' },
    });
    expect(Object.keys(buildCFNFileJSON?.Outputs)).toEqual(['snsTopicArn']);
    const meta = getProjectMeta(projRoot);
    const { snsTopicArn: customResourceSNSArn } = Object.keys(meta.custom).map((key) => meta.custom[key])[0].output;
    expect(customResourceSNSArn).toBeDefined();

    // using latest code, pull down the project
    const projRoot2 = await createNewProjectDir('customMigration2');
    const usingLatestCode = true;
    try {
      await amplifyPull(projRoot2, { emptyDir: true, appId }, usingLatestCode);
      assertNoParameterChangesBetweenProjects(projRoot, projRoot2);
      expect(collectCloudformationDiffBetweenProjects(projRoot, projRoot2)).toMatchSnapshot();
      await amplifyPushAuth(projRoot2, usingLatestCode);
      assertNoParameterChangesBetweenProjects(projRoot, projRoot2);
      expect(collectCloudformationDiffBetweenProjects(projRoot, projRoot2)).toMatchSnapshot();

      // building custom resources succeeds against a v10 cdk stack, even when using vLatest to build
      await expect(buildCustomResources(projRoot2, usingLatestCode)).resolves.not.toThrow();

      // migrate overrides to use vLatest
      const srcVLatestCustomResourceFilePath = path.join(__dirname, '..', '..', '..', 'custom-resources', 'custom-cdk-stack-vLatest.ts');
      const destVLatestCustomResourceFilePath = path.join(projRoot2, 'amplify', 'backend', 'custom', cdkResourceName, 'cdk-stack.ts');
      fs.copyFileSync(srcVLatestCustomResourceFilePath, destVLatestCustomResourceFilePath);

      // this should fail because customer also needs to update package.json dependencies for cdkV2
      await expect(buildCustomResources(projRoot2, usingLatestCode)).rejects.toThrow();

      // emulate updating the package.json dependencies
      const srcVLatestCustomPackageJSONFilePath = path.join(
        __dirname,
        '..',
        '..',
        '..',
        'custom-resources',
        'custom-cdk-stack-vLatest.package.json',
      );
      const destVLatestCustomPackageJSONFilePath = path.join(projRoot2, 'amplify', 'backend', 'custom', cdkResourceName, 'package.json');
      fs.copyFileSync(srcVLatestCustomPackageJSONFilePath, destVLatestCustomPackageJSONFilePath);

      // this should pass now
      await buildCustomResources(projRoot2, usingLatestCode);
      await amplifyPushAuth(projRoot2, usingLatestCode);

      // // Using latest code, add custom CFN and add dependency of custom CDK resource on the custom CFN
      await addCFNCustomResource(projRoot2, { name: cfnResourceName, promptForCustomResourcesSelection: true }, usingLatestCode);
      const customCFNFilePath = path.join(
        projRoot2,
        'amplify',
        'backend',
        'custom',
        cfnResourceName,
        `${cfnResourceName}-cloudformation-template.json`,
      );
      const customCFNFileJSON: any = JSONUtilities.readJson(customCFNFilePath);
      // Make sure input params has params from the resource dependency
      expect(customCFNFileJSON?.Parameters).toEqual({
        env: { Type: 'String' },
        [`custom${cdkResourceName}snsTopicArn`]: {
          Type: 'String',
          Description: `Input parameter describing snsTopicArn attribute for custom/${cdkResourceName} resource`,
        },
      });
    } finally {
      deleteProjectDir(projRoot2);
    }
  });
});
