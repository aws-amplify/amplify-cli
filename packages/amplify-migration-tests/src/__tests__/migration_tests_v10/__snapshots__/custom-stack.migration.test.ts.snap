// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adding custom resources migration test add/update CDK and CFN custom resources 1`] = `""`;

exports[`adding custom resources migration test add/update CDK and CFN custom resources 2`] = `""`;
